### ISSUE TITLE
[IO] Writing HistFactory model file twice gives strange results since ROOT 6.26.02

### ISSUE THREAD

**Issue by guitargeek** [root-project/root#12783]
## Problem description

Since 6.26.02, there is a problem when writing different HistFactory workspaces in the same script/executable. The second workspace file is corrupt, even if the only difference for the second time is the output file prefix.

That the second workspace is corrupt gets apparent in two ways:
1. It can't be opened with newer ROOT versions (but surprisingly it can be opened with the same 6.26 version)
2. The file sizes are much smaller

The first in commit in the 6.26 branch that shows this problem was found with git bisect (see the end of this comment).

I marked this issue as **high** priority, because writing out HistFactory workspaces is a common practice in ATLAS. And recently, several ATLAS workspaces unexplainably appear corrupt. Fixing this issue could fix these problems.

## To reproduce

1. Run the following script with ROOT 6.26.02 (or any newer 6.26 commit):
   ```c++
   {
      TFile f("myFile.root", "RECREATE");
      TH1D h("myHist", "", 10, 0, 10);
      h.Write();
   }
   std::vector<std::string> names = {"ws1", "ws2"};
   for (auto name : names) {
      RooStats::HistFactory::Measurement meas(name.c_str(), name.c_str());
      meas.SetOutputFilePrefix((name + "/WS").c_str());
      meas.SetPOI("mu");
      meas.SetLumi(1);
      meas.AddConstantParam("Lumi");
      meas.SetExportOnly(true);
      RooStats::HistFactory::Channel chan("SR");
      chan.SetData("", "");
      RooStats::HistFactory::Sample sample("samp");
      sample.SetHistoName("myHist");
      sample.SetInputFile("myFile.root");
      sample.AddNormFactor("mu", 1, 0, 100);
      chan.AddSample(sample);
      meas.AddChannel(chan);
      meas.CollectHistograms();
      //if(name == "ws2")
      //delete RooStats::HistFactory::MakeModelAndMeasurementFast(meas);
      RooStats::HistFactory::MakeModelAndMeasurementFast(meas);
   }
   ```
2. Validate the file sizes:
   ```bash
   ls -lah ws1 ws2
   ```
4. Try to open the second workspace with any ROOT 6.28 version:
   ```bash
   root ws2/WS_combined_ws2_model.root -b -q -e "combined->GetName()"
   ```

There will be errors such as:
```
Error in <TBufferFile::ReadClassBuffer>: Could not find the StreamerInfo for version 2 of the class RooAbsReal, object skipped at offset 6484979
Error in <TBufferFile::CheckByteCount>: object of class RooAbsReal read too few bytes: 2 instead of 336
```

## Information on the origin of the problem

The first bad commit in the 6.26 branch is 00ca5ca9b23f2 from the PR https://github.com/root-project/root/pull/10215.
The comes from a backport of https://github.com/root-project/root/pull/9486 to the 6.26 patch release branch.

Before that commit, when you run the reproducer script above, the `ws1` and `ws2` files are compatible:
```bash
ls -lah ws1 ws2
```
```txt
ws1:
total 88K
drwxr-xr-x 2 rembserj rembserj 4.0K May  7 23:31 .
drwxr-xr-x 4 rembserj rembserj 4.0K May  7 23:31 ..
-rw-r--r-- 1 rembserj rembserj  36K May  7 23:31 WS_combined_ws1_model.root
-rw-r--r-- 1 rembserj rembserj   14 May  7 23:31 WS_results.table
-rw-r--r-- 1 rembserj rembserj  34K May  7 23:31 WS_SR_ws1_model.root
-rw-r--r-- 1 rembserj rembserj  423 May  7 23:31 WS_ws1.root

ws2:
total 88K
drwxr-xr-x 2 rembserj rembserj 4.0K May  7 23:31 .
drwxr-xr-x 4 rembserj rembserj 4.0K May  7 23:31 ..
-rw-r--r-- 1 rembserj rembserj  36K May  7 23:31 WS_combined_ws2_model.root
-rw-r--r-- 1 rembserj rembserj   14 May  7 23:31 WS_results.table
-rw-r--r-- 1 rembserj rembserj  34K May  7 23:31 WS_SR_ws2_model.root
-rw-r--r-- 1 rembserj rembserj  423 May  7 23:31 WS_ws2.root
```

from 00ca5ca9b23f2 onward, the second workspace files are smaller from some reason:
```txt
ws1:
total 88K
drwxr-xr-x 2 rembserj rembserj 4.0K May  7 23:28 .
drwxr-xr-x 4 rembserj rembserj 4.0K May  7 23:28 ..
-rw-r--r-- 1 rembserj rembserj  36K May  7 23:28 WS_combined_ws1_model.root
-rw-r--r-- 1 rembserj rembserj   14 May  7 23:28 WS_results.table
-rw-r--r-- 1 rembserj rembserj  34K May  7 23:28 WS_SR_ws1_model.root
-rw-r--r-- 1 rembserj rembserj  423 May  7 23:28 WS_ws1.root

ws2:
total 56K
drwxr-xr-x 2 rembserj rembserj 4.0K May  7 23:28 .
drwxr-xr-x 4 rembserj rembserj 4.0K May  7 23:28 ..
-rw-r--r-- 1 rembserj rembserj  20K May  7 23:28 WS_combined_ws2_model.root
-rw-r--r-- 1 rembserj rembserj   14 May  7 23:28 WS_results.table
-rw-r--r-- 1 rembserj rembserj  19K May  7 23:28 WS_SR_ws2_model.root
-rw-r--r-- 1 rembserj rembserj  423 May  7 23:28 WS_ws2.root
```

So the problem somehow seems to be related to multithreaded IO?

**Comment by pcanal** [comment]
> So the problem somehow seems to be related to multithreaded IO?

Are you doing multi-threaded I/O in that example?

Note that the commit mentioned in the original post, helps with thread scalability but will effect serial code too.

The fact that before this commit, the reproducer works but reverting the commit on the master is not resolving the problem is strange, i.e. indicates some interaction between that commit and some other non-obvious commit.  (The other non obvious commit could be found by running a bisect where after each commit is checked out, one reverts the noted commits (according to the description this should lead to a working state up to the point where some other commit is make things 'worse' for this use case).

**Comment by pcanal** [comment]
This is fixed (independently/in two different ways) by both #12846 and #12848

**Comment by guitargeek** [comment]
Alright, I asked the RooFit ATLAS devs (in particular @will-cern who originally found this issue), and the conclusion was that we don't need a fixed release for the 6.26 series unless we get specific requests.
